File: backend/app/models/analysis.py
"""挥杆分析相关模型（对齐 docs/03-数据库设计文档.md 3.2-3.4 节）."""

from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin


class AnalysisQuota(Base, TimestampMixin):
    """月度分析配额."""

    __tablename__ = "analysis_quotas"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    quota_month: Mapped[str] = mapped_column(String(7), nullable=False)  # 2026-04
    used: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    total: Mapped[int] = mapped_column(Integer, default=3, server_default="3")  # -1 = 无限
    bonus: Mapped[int] = mapped_column(Integer, default=0, server_default="0")

    __table_args__ = (
        UniqueConstraint("user_id", "quota_month", name="uq_analysis_quota"),
        Index("idx_analysis_quotas_lookup", "user_id", "quota_month"),
    )


class SwingAnalysis(Base, TimestampMixin):
    """挥杆分析记录."""

    __tablename__ = "swing_analyses"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )

    # 输入
    video_url: Mapped[str] = mapped_column(String(512), nullable=False)
    video_duration: Mapped[Decimal | None] = mapped_column(Numeric(5, 2), nullable=True)
    video_file_size: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    camera_angle: Mapped[str] = mapped_column(String(20), nullable=False)  # face_on / down_the_line
    club_type: Mapped[str] = mapped_column(String(20), nullable=False)
    analysis_mode: Mapped[str] = mapped_column(
        String(20), nullable=False, default="full_swing", server_default="'full_swing'"
    )
    # M10-03：用户声明的目标码数（yard），供 yardage book 历史反推
    target_yardage: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # M7-14：评分管线版本标记。
    # 不做"二次评分"。
    engine_version: Mapped[str] = mapped_column(
        String(20), default="v1", server_default="'v1'", nullable=False
    )

    # 状态
    status: Mapped[str] = mapped_column(String(20), default="pending", server_default="'pending'")
    stage: Mapped[str | None] = mapped_column(String(30), nullable=True)
    stage_progress: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    error_code: Mapped[int | None] = mapped_column(Integer, nullable=True)
    error_message: Mapped[str | None] = mapped_column(String(500), nullable=True)
    quota_refunded: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    priority: Mapped[str] = mapped_column(String(10), default="standard", server_default="'standard'")

    # 输出
    overall_score: Mapped[int | None] = mapped_column(Integer, nullable=True)
    score_change: Mapped[int | None] = mapped_column(Integer, nullable=True)
    phase_scores: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # M10-01/02：推杆/切杆 mode 专属特征分（pendulum_stability 等）；full_swing 为 NULL
    mode_feature_scores: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    phase_timestamps: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    skeleton_video_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    skeleton_data_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    thumbnail_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    share_card_url: Mapped[str | None] = mapped_column(String(512), nullable=True)

    # 引擎非阻断质量提示（JSON 字符串列表），如 ["low_light","camera_shake"]
    quality_warnings: Mapped[list | None] = mapped_column(JSONB, nullable=True)

    # P2-M7-06：整体置信度 0-1（V1 引擎报告兜底 1.0；客户端 <0.5 展示「建议重拍」CTA）
    analysis_confidence: Mapped[float] = mapped_column(
        Float, nullable=False, default=1.0, server_default="1.0"
    )
    # P2-M7-06：每特征 confidence dict (feature_name → 0-1)；V1 引擎为 {}
    feature_confidences: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # P2-W10：W8 引擎诊断（codec/HDR/慢动作/fps/audio/fallback），客户端调试浮层展示
    # 结构：list[{code, level, detail?, ts}]；V1 引擎或老报告为 NULL
    engine_warnings: Mapped[list | None] = mapped_column(JSONB, nullable=True)

    # 是否为示例视频体验（不计入配额、不入历史）
    is_sample: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")

    analyzed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # 用户侧软删除：仅隐藏，不物理删；保留用于审计/溯源
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # 关系
    user = relationship("User", back_populates="analyses", lazy="noload")
    issues: Mapped[list["AnalysisIssue"]] = relationship(
        back_populates="analysis",
        cascade="all, delete-orphan",
        lazy="noload",
    )
    recommendations: Mapped[list["AnalysisRecommendation"]] = relationship(
        back_populates="analysis",
        cascade="all, delete-orphan",
        lazy="noload",
    )

    __table_args__ = (
        CheckConstraint(
            "analysis_mode IN ('full_swing', 'putting', 'chipping')",
            name="chk_swing_analysis_mode",
        ),
        Index("idx_swing_analyses_user", "user_id", "created_at"),
        Index("idx_swing_analyses_status", "status"),
        Index("idx_swing_analyses_confidence", "analysis_confidence"),
        Index("idx_swing_analyses_engine_version", "engine_version"),
    )


class AnalysisIssue(Base, TimestampMixin):
    """分析报告中的问题诊断项."""

    __tablename__ = "analysis_issues"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    analysis_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("swing_analyses.id", ondelete="CASCADE"),
        nullable=False,
    )
    issue_type: Mapped[str] = mapped_column(String(40), nullable=False)  # casting / over_the_top ...
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    severity: Mapped[str] = mapped_column(String(10), nullable=False)  # high / medium / low
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    key_frame_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    key_frame_timestamp: Mapped[Decimal | None] = mapped_column(Numeric(5, 2), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    # P2-M7-06：每诊断 confidence + tier（V1 引擎为 NULL）
    confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    confidence_tier: Mapped[str | None] = mapped_column(String(20), nullable=True)

    analysis = relationship("SwingAnalysis", back_populates="issues", lazy="noload")

    __table_args__ = (
        Index("idx_analysis_issues_analysis", "analysis_id"),
    )


class AnalysisRecommendation(Base, TimestampMixin):
    """分析报告中的训练建议项."""

    __tablename__ = "analysis_recommendations"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    analysis_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("swing_analyses.id", ondelete="CASCADE"),
        nullable=False,
    )
    drill_id: Mapped[str] = mapped_column(String(40), nullable=False)
    target_issue: Mapped[str | None] = mapped_column(String(40), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")

    analysis = relationship("SwingAnalysis", back_populates="recommendations", lazy="noload")

    __table_args__ = (
        Index("idx_analysis_recs_analysis", "analysis_id"),
    )

File: app/lib/data/models/analysis.dart
// 挥杆分析 DTO：对照 client/src/types/analysis.ts，与 backend schemas/analysis.py 对齐。

class UploadToken {
  final String uploadId;
  final String uploadUrl;
  final String key;
  final Map<String, String> fields;
  final String? expiresAt;

  const UploadToken({
    required this.uploadId,
    required this.uploadUrl,
    required this.key,
    required this.fields,
    this.expiresAt,
  });

  factory UploadToken.fromJson(Map<String, dynamic> j) => UploadToken(
        uploadId: j['upload_id']?.toString() ?? '',
        uploadUrl: j['upload_url']?.toString() ?? '',
        key: j['key']?.toString() ?? '',
        fields: (j['fields'] as Map?)?.map(
              (k, v) => MapEntry(k.toString(), v.toString()),
            ) ??
            const {},
        expiresAt: j['expires_at'] as String?,
      );
}

class CreateAnalysisResult {
  final String analysisId;
  final String status;
  final int queuePosition;
  final int estimatedSeconds;

  const CreateAnalysisResult({
    required this.analysisId,
    required this.status,
    required this.queuePosition,
    required this.estimatedSeconds,
  });

  factory CreateAnalysisResult.fromJson(Map<String, dynamic> j) =>
      CreateAnalysisResult(
        analysisId: j['analysis_id']?.toString() ?? '',
        status: j['status']?.toString() ?? 'pending',
        queuePosition: (j['queue_position'] as num?)?.toInt() ?? 0,
        estimatedSeconds: (j['estimated_seconds'] as num?)?.toInt() ?? 25,
      );
}

class AnalysisErrorInfo {
  final int code;
  final String message;
  final bool quotaRefunded;
  const AnalysisErrorInfo(
      {required this.code, required this.message, this.quotaRefunded = false});

  factory AnalysisErrorInfo.fromJson(Map<String, dynamic> j) => AnalysisErrorInfo(
        code: (j['code'] as num?)?.toInt() ?? 0,
        message: j['message']?.toString() ?? '',
        quotaRefunded: j['quota_refunded'] == true,
      );
}

class AnalysisStatusInfo {
  final String analysisId;
  final String status; // pending/processing/completed/failed
  final String? stage;
  final int stageProgress;
  final int? estimatedRemainingSeconds;
  final AnalysisErrorInfo? error;

  const AnalysisStatusInfo({
    required this.analysisId,
    required this.status,
    this.stage,
    this.stageProgress = 0,
    this.estimatedRemainingSeconds,
    this.error,
  });

  bool get isCompleted => status == 'completed';
  bool get isFailed => status == 'failed';

  factory AnalysisStatusInfo.fromJson(Map<String, dynamic> j) =>
      AnalysisStatusInfo(
        analysisId: j['analysis_id']?.toString() ?? '',
        status: j['status']?.toString() ?? 'pending',
        stage: j['stage'] as String?,
        stageProgress: (j['stage_progress'] as num?)?.toInt() ?? 0,
        estimatedRemainingSeconds:
            (j['estimated_remaining_seconds'] as num?)?.toInt(),
        error: j['error'] is Map<String, dynamic>
            ? AnalysisErrorInfo.fromJson(j['error'] as Map<String, dynamic>)
            : null,
      );
}

class PhaseScore {
  final num score;
  final String label;
  final bool isWeakest;
  const PhaseScore(
      {required this.score, required this.label, this.isWeakest = false});

  factory PhaseScore.fromJson(Map<String, dynamic> j) => PhaseScore(
        score: (j['score'] as num?) ?? 0,
        label: j['label']?.toString() ?? '',
        isWeakest: j['is_weakest'] == true,
      );
}

class PhaseWindow {
  final double start;
  final double end;
  const PhaseWindow({required this.start, required this.end});

  factory PhaseWindow.fromJson(Map<String, dynamic> j) => PhaseWindow(
        start: (j['start'] as num?)?.toDouble() ?? 0,
        end: (j['end'] as num?)?.toDouble() ?? 0,
      );
}

class AnalysisIssue {
  final String type;
  final String name;
  final String severity; // high/medium/low
  final String description;
  final String? keyFrameUrl;
  final double? keyFrameTimestamp;
  final String? confidenceTier; // confirmed/leaning/hidden

  const AnalysisIssue({
    required this.type,
    required this.name,
    required this.severity,
    required this.description,
    this.keyFrameUrl,
    this.keyFrameTimestamp,
    this.confidenceTier,
  });

  factory AnalysisIssue.fromJson(Map<String, dynamic> j) => AnalysisIssue(
        type: j['type']?.toString() ?? '',
        name: j['name']?.toString() ?? '',
        severity: j['severity']?.toString() ?? 'medium',
        description: j['description']?.toString() ?? '',
        keyFrameUrl: j['key_frame_url'] as String?,
        keyFrameTimestamp: (j['key_frame_timestamp'] as num?)?.toDouble(),
        confidenceTier: j['confidence_tier'] as String?,
      );
}

class AnalysisRecommendation {
  final String drillId;
  final String? targetIssue;
  final int sortOrder;
  const AnalysisRecommendation(
      {required this.drillId, this.targetIssue, this.sortOrder = 0});

  factory AnalysisRecommendation.fromJson(Map<String, dynamic> j) =>
      AnalysisRecommendation(
        drillId: j['drill_id']?.toString() ?? '',
        targetIssue: j['target_issue'] as String?,
        sortOrder: (j['sort_order'] as num?)?.toInt() ?? 0,
      );
}

class AnalysisReport {
  final String id;
  final String status;
  final String cameraAngle;
  final String clubType;
  final String? analysisMode;
  final String videoUrl;
  final double? videoDuration;
  final String? skeletonVideoUrl;
  final String? thumbnailUrl;
  final num? overallScore;
  final String? scoreLevel;
  final num? scoreChange;
  final Map<String, PhaseScore> phaseScores;
  final Map<String, PhaseWindow> phaseTimestamps;
  final List<AnalysisIssue> issues;
  final List<AnalysisRecommendation> recommendations;
  final List<String> qualityWarnings;
  final num? analysisConfidence;
  final List<String> phaseHighlights;
  final String? createdAt;
  final String? analyzedAt;

  const AnalysisReport({
    required this.id,
    required this.status,
    required this.cameraAngle,
    required this.clubType,
    this.analysisMode,
    required this.videoUrl,
    this.videoDuration,
    this.skeletonVideoUrl,
    this.thumbnailUrl,
    this.overallScore,
    this.scoreLevel,
    this.scoreChange,
    this.phaseScores = const {},
    this.phaseTimestamps = const {},
    this.issues = const [],
    this.recommendations = const [],
    this.qualityWarnings = const [],
    this.analysisConfidence,
    this.phaseHighlights = const [],
    this.createdAt,
    this.analyzedAt,
  });

  factory AnalysisReport.fromJson(Map<String, dynamic> j) => AnalysisReport(
        id: j['id']?.toString() ?? '',
        status: j['status']?.toString() ?? 'completed',
        cameraAngle: j['camera_angle']?.toString() ?? 'face_on',
        clubType: j['club_type']?.toString() ?? 'unknown',
        analysisMode: j['analysis_mode'] as String?,
        videoUrl: j['video_url']?.toString() ?? '',
        videoDuration: (j['video_duration'] as num?)?.toDouble(),
        skeletonVideoUrl: j['skeleton_video_url'] as String?,
        thumbnailUrl: j['thumbnail_url'] as String?,
        overallScore: j['overall_score'] as num?,
        scoreLevel: j['score_level'] as String?,
        scoreChange: j['score_change'] as num?,
        phaseScores: (j['phase_scores'] as Map?)?.map(
              (k, v) => MapEntry(
                  k.toString(), PhaseScore.fromJson(v as Map<String, dynamic>)),
            ) ??
            const {},
        phaseTimestamps: (j['phase_timestamps'] as Map?)?.map(
              (k, v) => MapEntry(k.toString(),
                  PhaseWindow.fromJson((v as Map).cast<String, dynamic>())),
            ) ??
            const {},
        issues: (j['issues'] as List?)
                ?.map((e) => AnalysisIssue.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        recommendations: (j['recommendations'] as List?)
                ?.map((e) =>
                    AnalysisRecommendation.fromJson(e as Map<String, dynamic>))
                .toList() ??
            const [],
        qualityWarnings: (j['quality_warnings'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        analysisConfidence: j['analysis_confidence'] as num?,
        phaseHighlights: (j['phase_highlights'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        createdAt: j['created_at'] as String?,
        analyzedAt: j['analyzed_at'] as String?,
      );
}

class AnalysisListPage {
  final List<AnalysisListItem> items;
  final int total;
  final int page;
  final int pageSize;
  const AnalysisListPage({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });
}

class AnalysisListItem {
  final String id;
  final String status;
  final String clubType;
  final String cameraAngle;
  final String? thumbnailUrl;
  final num? overallScore;
  final String? scoreLevel;
  final num? scoreChange;
  final String? analyzedAt;
  final String? createdAt;

  const AnalysisListItem({
    required this.id,
    required this.status,
    required this.clubType,
    required this.cameraAngle,
    this.thumbnailUrl,
    this.overallScore,
    this.scoreLevel,
    this.scoreChange,
    this.analyzedAt,
    this.createdAt,
  });

  factory AnalysisListItem.fromJson(Map<String, dynamic> j) => AnalysisListItem(
        id: j['id']?.toString() ?? '',
        status: j['status']?.toString() ?? 'completed',
        clubType: j['club_type']?.toString() ?? 'unknown',
        cameraAngle: j['camera_angle']?.toString() ?? 'face_on',
        thumbnailUrl: j['thumbnail_url'] as String?,
        overallScore: j['overall_score'] as num?,
        scoreLevel: j['score_level'] as String?,
        scoreChange: j['score_change'] as num?,
        analyzedAt: j['analyzed_at'] as String?,
        createdAt: j['created_at'] as String?,
      );
}
